invoices/invoiceregistry.go

```diff
@@ -106,7 +106,10 @@ func NewRegistry(cdb *channeldb.DB, expiryWatcher *InvoiceExpiryWatcher,
 
 // Start starts the registry and all goroutines it needs to carry out its task.
 func (i *InvoiceRegistry) Start() error {
-	i.expiryWatcher.Start(i.CancelInvoice)
+	if err := i.expiryWatcher.Start(i.CancelInvoice); err != nil {
```

Comment:
is this commit a workaround in case `InvoiceRegistry.Start` is called twice? why not add `sync.Once` pattern here and for Stop instead?

Reply:
I think I suggested in an earlier revision to use the (atomic) `started`/`stopped` pattern. I think it is better to error when `Start` or `Stop` are called multiple times, to catch possible bugs.

Reply:
Initially I used `sync.Once` but following reviewers comments decided to rework it. This intends to let either `Start` or `Stop` running, but let `Stop` wait for `Start` to finish if called "early". 

As for whether it's better to error:
While `Start` will error if called multiple times, imho it's better to let `Stop` be called multiple times in the in a start/stop situation as multiple objects may hold a reference to the instance that needs to be stopped, but this way only one will actually stop, all the others will fall trough (without error).

Reply:
> to catch possible bugs.

what bugs are we preventing exactly? it's a singleton object, we don't care if it is started or stopped multiple times. the important thing is that any callers block until the operation has been completed

Reply:
>  let Stop be called multiple times in the in a start/stop situation as multiple objects may hold a reference to the instance that needs to be stopped, but this way only one will actually stop, all the others will fall trough (without error).

I agree here, stopping should definitely support multiple calls without failure

Reply:
I think in the way we use it, it should never be started multiple times. Maybe it is called from a different function that itself is really supposed to be called only once. In that case, an error here would expose that. All solutions work, so maybe it is mostly academical. I like the opportunity to catch those things.

Reply:
i understand that we don't expect it to be called multiple times, but there is no harm in trying to start something that has already been started. so long as the resource is ready, the caller can begin using the api. it's incredibly unlikely that we will accidentally start the invoice twice in production, and even if we do, failing doesn't actually prevent any unintended consequences. all of this doesn't seem worth it to fix an benign, unlikely bug that only results in more code complexity and lower test coverage.

Reply:
>it's incredibly unlikely that we will accidentally start the invoice twice in production

Still those things happen in software projects in general. all of a sudden there are two instances of the same thing.

What failing does is making us developers aware of a potential bug instead of silently ignoring an unexpected sequence of events.

To me, the atomic add or the sync.once isn't very different in terms of complexity.

But this is totally non-blocking. I am fine with sync.once too.

Reply:
maybe it is just a difference in philosophy. my opinion is that if we have the ability to make the API robust against user error in a way which doesn't have negative consequences we should take that opportunity. the fact that a user can take the object in an unknown state, call start, and safely begin using is a feature, not bug imo :)

Reply:
So to summarize `InvoiceExpiryWatcher.Start()` can be called only once, and will return an error otherwise (unless stopped), while `InvoiceExpiryWatcher.Stop()` can be called any times and the first one will stop the watcher.

I personally believe this is a better approach than the compare and exchange or the `sync.Once` solution  as it's more flexible but happy to revert this in favor of the one we use in the code if needed.